proposals/safe-fixed-size-buffers.md

```diff
@@ -0,0 +1,378 @@
+
+Attempting to assign to an element of a readonly fixed-size buffer or pass it as an `out` or `ref` parameter in any other context is a compile-time error.
+This is achieved by the following.
+
+A member access for a readonly fixed-size buffer is evaluated and classified as follows:
+
+- If access occurs in a context where direct assignments to an element of readonly fixed-size buffer are permitted, the result of the expression is classified as a value of type `System.Span<S>`, where S is the element type of the fixed-size buffer.
+  The value can be used to access members’ elements.
+- Otherwise, the expression is classified as a value of type `System.ReadOnlySpan<S>`, where S is the element type of the fixed-size buffer.
+  The value can be used to access members’ elements.
+
+### Definite assignment checking
+
+Fixed-size buffers are not subject to definite assignment-checking, and fixed-size buffer members are ignored for purposes of definite-assignment checking of struct type variables.
+
+When a fixed-size buffer member is static or the outermost containing struct variable of a fixed-size buffer member is a static variable, an instance variable of a class instance, or an array element, the elements of the fixed-size buffer are automatically initialized to their default values. In all other cases, the initial content of a fixed-size buffer is undefined.
+
+### Metadata
+
+#### Metadata emit and code generation
+
+For metadata encoding compiler will rely on recently added [```System.Runtime.CompilerServices.InlineArrayAttribute```](https://github.com/dotnet/runtime/issues/61135). 
+
+Fixed-size buffers like:
+``` C#
+public partial class C
+{
+    public int buffer1[10];
+    public readonly int buffer2[10];
+}
+```
+will be emitted as  fields of a specially decorated struct type.
+
+Equivalent C# code will be:
+
+``` C#
+public partial class C
+{
+    public Buffer10<int> buffer1;
+    public readonly Buffer10<int> buffer2;
+}
+
+[System.Runtime.CompilerServices.InlineArray(10)]
+public struct Buffer10<T>
+{
+    private T _element0;
+
+    [UnscopedRef]
+    public System.Span<T> AsSpan()
+    {
+        return System.Runtime.InteropServices.MemoryMarshal.CreateSpan(ref _element0, 10);
+    }
+
+    [UnscopedRef]
+    public readonly System.ReadOnlySpan<T> AsReadOnlySpan()
+    {
+        // Note, an attempt to compile reports error CS1605: Cannot use '_element0' as a ref or out value because it is read-only
+        return System.Runtime.InteropServices.MemoryMarshal.CreateReadOnlySpan(ref _element0, 10);  
+    }
+}
+```
+
+The actual naming conventions for the type and its members are TBD. The framework will likely include a set of predefined "buffer" types that cover
+a limited set of buffer sizes. When a predefined type doesn't exist, compiler will synthesize it in the module being built. Names of the generated types
+will be "speakable" in order to support consumption from other languages. 
```

Comment:
Agree they should be speakable. I was thinking about what the best namespace would be for these types when the compiler generates them. Ended up thinking that `System.Runtime.CompilerServices` is a good candidate. Mostly cause that is the likely place the libraries team would want to put the predefined types. That would mean 

1. The libraries team could use the compiler to effectively generate the types they need vs. hand coding them since we're putting them in the right place.
2. It's a consistent location for the types to always exist in

My other thought was `Microsoft.CodeAnalysis`.

Reply:
> Mostly cause that is the likely place the libraries team would want to put the predefined types.

From the BCL point of view, I do not think we want the compiler to ever auto-generate types in the `System.Runtime.CompilerServices` namespace. That's would be a factory for type-name collisions.

Can we make the physicial type of the field a hidden internal implementation detail and only expose these fields publicly via the Span and ReadOnlySpan accessors?

Reply:
> Can we make the physicial type of the field a hidden internal implementation detail and only expose these fields publicly via the Span and ReadOnlySpan accessors?

This is my preference as well.  Basically mirror what we do today for unsafe fixed buffers: a developer writes `fixed int buffer[10]`, the compiler generates an unspeakable type for that, and the effective type visible to consuming code is `int*`.  For safe buffers, it could be the same approach: the developer writes `int buffer[10]`, the compiler generates an unspeakable type for that, and the effective type visible to consuming code is either `Span<int>` or `ReadOnlySpan<int>` depending on whether the field is readonly.

> The libraries team could use the compiler to effectively generate the types they need vs. hand coding them since we're putting them in the right place.

Handcoding these is trivial. _If_ we wanted to expose any publicly for size/sharing reasons, it would be easy to do so without the compiler feature.  As noted earlier, I'd also be concerned about using the compiler feature to implement these, as it would either be limiting for what the compiler could emit on them in the future (any additions the compiler made here would implicitly impact core library public surface area), or if we didn't want to constrain the compiler in that fashion (I don't want to), we wouldn't use it for this, anyway.